omnipaxos/src/sequence_paxos/leader.rs

```diff
@@ -63,14 +63,18 @@ where
                 });
             }
         } else {
-            self.state.0 = Role::Follower;
+            self.become_follower();
         }
     }
 
-    pub(crate) fn handle_preparereq(&mut self, from: NodeId) {
+    pub(crate) fn become_follower(&mut self) {
+        self.state.0 = Role::Follower;
+    }
+
+    pub(crate) fn handle_preparereq(&mut self, prepreq: PrepareReq, from: NodeId) {
         #[cfg(feature = "logging")]
         debug!(self.logger, "Incoming message PrepareReq from {}", from);
-        if self.state.0 == Role::Leader {
+        if self.state.0 == Role::Leader && prepreq.n <= self.leader_state.n_leader {
             self.leader_state.set_decided_idx(from, None);
```

Comment:
We need to reset the PromiseState of this node to be `NotPromised` as well here. Otherwise we will keep sending accepts to it since the `get_promised_followers()` is now based on the `leader_state.promises_meta` rather than `leader_state.decided_indexes`.

Looking at the code now, I'm not sure why we even have the `leader_state.decided_indexes`. The decided index should probably be stored in the `PromiseMetaData`. @kevin-harrison could you please double check if there is any reason to keep the decided_indexes separate? And if not then add this change?

Reply:
We take the max of `leader_state.decided_indexes` in `handle_majority_promises()` to sync with the up-to-date decided index after attaining leadership. If its possible for a node to have the greatest decided_idx but not the greatest `PromiseMetaData` then it makes since to have these two separate.

Reply:
right, but then let's anyway make the change and store the decided index in `PromiseMetaData`. We just need to make sure that the comparator for PromiseMetaData doesn't include the decided index, and we also need a new function `get_max_decided_idx()` that takes the max decided index among PromiseMetaData.